src/puppetlabs/trapperkeeper/services/webserver/jetty9_websockets.clj

```diff
@@ -0,0 +1,105 @@
+(ns puppetlabs.trapperkeeper.services.webserver.jetty9-websockets
+  (:import (clojure.lang IFn)
+           (org.eclipse.jetty.server Request)
+           (org.eclipse.jetty.websocket.api WebSocketAdapter Session)
+           (org.eclipse.jetty.websocket.server WebSocketHandler)
+           (org.eclipse.jetty.websocket.servlet WebSocketServletFactory WebSocketCreator)
+           (java.security.cert X509Certificate)
+           (java.nio ByteBuffer))
+
+  (:require [clojure.tools.logging :as log]
+            [puppetlabs.websockets.client :refer [WebSocketProtocol]]
+            [schema.core :as schema]))
+
+(def WebsocketHandlers
+  {(schema/optional-key :on-connect) IFn
+   (schema/optional-key :on-error) IFn
+   (schema/optional-key :on-close) IFn
+   (schema/optional-key :on-text) IFn
+   (schema/optional-key :on-bytes) IFn})
+
+(defprotocol WebSocketSend
+  (-send! [x ws] "How to encode content sent to the WebSocket clients"))
+
+(extend-protocol WebSocketSend
+  (Class/forName "[B")
+  (-send! [ba ws]
+    (-send! (ByteBuffer/wrap ba) ws))
+
+  ByteBuffer
+  (-send! [bb ws]
+    (-> ^WebSocketAdapter ws .getRemote (.sendBytes ^ByteBuffer bb)))
+
+  String
+  (-send! [s ws]
+    (-> ^WebSocketAdapter ws .getRemote (.sendString ^String s))))
+
+(extend-protocol WebSocketProtocol
+  WebSocketAdapter
+  (send! [this msg]
+    (-send! msg this))
+  (close! [this]
+    (.. this (getSession) (close)))
+  (remote-addr [this]
+    (.. this (getSession) (getRemoteAddress)))
+  (ssl? [this]
+    (.. this (getSession) (getUpgradeRequest) (isSecure)))
+  (peer-certs [this]
+    (.. this (getCerts)))
+  (idle-timeout! [this ms]
+    (.. this (getSession) (setIdleTimeout ^long ms)))
+  (connected? [this]
+    (. this (isConnected))))
+
+(defn- do-nothing [& args])
+
+(definterface CertGetter
+  (^Object getCerts []))
+
+(schema/defn ^:always-validate proxy-ws-adapter :- WebSocketAdapter
+  [handlers :- WebsocketHandlers
+   x509certs :- [X509Certificate]]
+  (let [{:keys [on-connect on-error on-text on-close on-bytes]
+         :or {on-connect do-nothing
+              on-error do-nothing
+              on-text do-nothing
+              on-close do-nothing
+              on-bytes do-nothing}} handlers]
+    (proxy [WebSocketAdapter CertGetter] []
+      (onWebSocketConnect [^Session session]
+        (let [^WebSocketAdapter this this]
+          (proxy-super onWebSocketConnect session))
+        (on-connect this))
+      (onWebSocketError [^Throwable e]
+        (on-error this e))
+      (onWebSocketText [^String message]
+        (on-text this message))
+      (onWebSocketClose [statusCode ^String reason]
+        (let [^WebSocketAdapter this this]
+          (proxy-super onWebSocketClose statusCode reason))
```

Comment:
Hmm, I could go either way on this.

I was thinking it would be beneficial to call into them in case future versions of Jetty change the `WebsocketAdapter` implementation of those methods and we ever want to upgrade. We ran into an issue similar to that a while ago, wherein we were overriding a method in Jetty's `ProxyServlet` class with `proxy`, but the implementation of that method changed in a later version of Jetty. This was causing a bunch of issues with our tests when we tried to upgrade, and it took us a long time to figure out that the Jetty-side implementation of the method we were overriding had changed, and that we had to change our implementation of the method as well. I think @camlow325 did that work, so he might have some more insight.

Saying that, while I have very little knowledge of Websockets, my guess is that the implementation of those methods is unlikely to change, in which case I think it would be fine to not call into them.

I think I'd lean towards calling into them for the extra safety, but I don't have a super strong opinion either way. @richardc @cprice404 what do you think?